core/primitives/src/cpu.rs

```diff
@@ -0,0 +1,158 @@
+#[derive(Debug)]
+pub enum Error {
+    /// Unsupported architecture
+    Architecture,
+    /// The following required features are not supported by the CPU.
+    MissingFeatures(Vec<&'static str>),
+    /// SSE is available but disabled.
+    SSEDisabled,
+    /// AVX is available but disabled.
+    AVXDisabled,
+}
+
+impl Error {
+    pub const NOTE: &'static str = "confirm hardware requirements at \
+        <https://docs.near.org/docs/develop/node/validator/hardware>.";
+
+    pub fn terminate(&self) -> ! {
+        eprintln!("error: {}\nnote: {}", self, Self::NOTE);
+        std::process::exit(1);
+    }
+}
+
+impl std::error::Error for Error {}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use Error::*;
+        match self {
+            Architecture => f.write_str("unsupported CPU architecture")?,
+            SSEDisabled => f.write_str("SSE CPU feature is supported but disabled")?,
+            AVXDisabled => f.write_str("AVX CPU feature is supported but disabled")?,
+            MissingFeatures(features) => {
+                f.write_str("required CPU features are missing: ")?;
+                let mut features = features.iter().peekable();
+                while let Some(feature) = features.next() {
+                    f.write_str(feature)?;
+                    if features.peek().is_some() {
+                        f.write_str(", ")?;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Try to ensure that the required CPU features are available and enabled for the current CPU.
+///
+/// If any of the features are not available, the returned error will specify which ones are
+/// missing.
+///
+/// Note that this is a best-effort kind of thing. If these features are already enabled when
+/// compiling the program, then we may very well end up executing some unsupported instruction
+/// before we even get to executing this code.
+pub fn verify_cpu_compatibility() -> Result<(), Error> {
+    imp::verify_cpu_compatibility()
+}
+
+mod imp {
```

Comment:
It looks like you want:

```rust
#[cfg(not(target_arch = "x86_64"))]
mod impl {
   // ...
}

#[cfg(target_arch = "x86_64")]
mod impl {
   // ...
}
```

and then the comment about CPUID being unsupported by some
architectures is moot.